manager/controllers/scaling/metrics_cache.go

```diff
@@ -0,0 +1,217 @@
+package scaling
+
+import (
+	"container/list"
+	"context"
+	"crypto/tls"
+	"fmt"
+	"net/http"
+	"sync"
+	"time"
+
+	dfv1 "github.com/argoproj-labs/argo-dataflow/api/v1alpha1"
+	lru "github.com/hashicorp/golang-lru"
+	pmodel "github.com/prometheus/client_model/go"
+	"github.com/prometheus/common/expfmt"
+	"k8s.io/apimachinery/pkg/util/runtime"
+	"k8s.io/client-go/tools/cache"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+var (
+	httpClient   *http.Client
+	metricsCache *lru.Cache
+)
+
+func init() {
+	t := http.DefaultTransport.(*http.Transport).Clone()
+	t.MaxIdleConns = 32
+	t.MaxConnsPerHost = 32
+	t.MaxIdleConnsPerHost = 32
+	t.TLSClientConfig = &tls.Config{InsecureSkipVerify: true}
+	httpClient = &http.Client{Timeout: 1 * time.Second, Transport: t}
+
+	metricsCache, _ = lru.New(1000)
+}
+
+type MetricsCacheHandler struct {
+	client client.Client
+
+	stepMap  map[string]*list.Element
+	stepList *list.List
+	lock     sync.RWMutex
+	workers  int
+}
+
+func NewMetricsCacheHandler(client client.Client, workers int) *MetricsCacheHandler {
+	return &MetricsCacheHandler{
+		client:  client,
+		workers: workers,
+
+		stepMap:  make(map[string]*list.Element),
+		stepList: list.New(),
+		lock:     sync.RWMutex{},
+	}
+}
+
+func (m *MetricsCacheHandler) Contains(step *dfv1.Step) (bool, error) {
+	key, err := cache.MetaNamespaceKeyFunc(step)
+	if err != nil {
+		return false, fmt.Errorf("failed to get key for step object: %w", err)
+	}
+	m.lock.RLock()
+	defer m.lock.RUnlock()
+	_, ok := m.stepMap[key]
+	return ok, nil
+}
+
+func (m *MetricsCacheHandler) Length() int {
+	m.lock.RLock()
+	defer m.lock.RUnlock()
+	return m.stepList.Len()
+}
+
+func (m *MetricsCacheHandler) StartWatchingStep(step *dfv1.Step) error {
+	key, err := cache.MetaNamespaceKeyFunc(step)
+	if err != nil {
+		return err
+	}
+	m.lock.Lock()
+	defer m.lock.Unlock()
+	if _, ok := m.stepMap[key]; !ok {
+		m.stepMap[key] = m.stepList.PushBack(step)
+	}
+	return nil
+}
+
+func (m *MetricsCacheHandler) StopWatchingStep(step *dfv1.Step) error {
+	key, err := cache.MetaNamespaceKeyFunc(step)
+	if err != nil {
+		return err
+	}
+	m.lock.Lock()
+	defer m.lock.Unlock()
+	if e, ok := m.stepMap[key]; ok {
+		_ = m.stepList.Remove(e)
+		delete(m.stepMap, key)
+	}
+	return nil
+}
+
+func (m *MetricsCacheHandler) Start(ctx context.Context) {
+	stepCh := make(chan *dfv1.Step)
+	ctx, cancel := context.WithCancel(ctx)
+	defer cancel()
+	for i := 1; i <= m.workers; i++ {
+		go m.pullMetrics(ctx, i, stepCh)
+	}
+
+	assign := func() {
+		m.lock.Lock()
+		defer m.lock.Unlock()
+		if m.stepList.Len() == 0 {
+			return
+		}
+		e := m.stepList.Front()
+		if step, ok := e.Value.(*dfv1.Step); ok {
+			m.stepList.MoveToBack(e)
+			stepCh <- step
+		}
+	}
+
+	for {
```

Comment:
I think this loop should use a ticker every 60s (configurable) to queue every step

Reply:
need to think about adding jitter to that, to avoid thundering herd

Reply:
how do you guarantee that we don't pull metrics too often?

Reply:
20 seconds is the max interval for a target metrics endpoint.

```
time.Sleep(time.Millisecond * time.Duration(func() int {
			l := m.Length()
			if l == 0 {
				return 20000
			}
			result := 20000 / l
			if result > 0 {
				return result
			}
			return 1
		}()))
```

Reply:
Explain a little bit on this:

Case 1: A in the queue, it sleeps 20s;
Case 2: [A, B] in the queue
      - A assigned, the queue becomes [B, A], sleep 10s, 
      - B assigned, the queue becomes [A, B] again, sleeps 10s.
 Case 3: [A, B, C] in the queue
      - A assigned, [A, B, C] - > [B, C, A], sleep 20/3 seconds;
      - B assigned, [B, C, A] - > [C, B, A], sleep 20/3 seconds;
      - C assigned, -> [A, B, C], sleep 20/3 seconds.

Reply:
I'm going to trust you. I can't think of an obvious way to do this, so this will do for now.